src/io.rs

```diff
@@ -0,0 +1,95 @@
+use std::io::{Error, Read, Result, Write};
+
+use libc::{c_void, size_t};
+
+use crate::error::rustls_io_error;
+
+/// A callback for rustls_server_session_read_tls or rustls_client_session_read_tls.
+/// An implementation of this callback should attempt to read up to n bytes from the
+/// network, storing them in `buf`. If any bytes were stored, the implementation should
+/// set out_n to the number of bytes stored and return 0. If there was an error,
+/// (including EAGAIN or EWOULDBLOCK), the implementation should return `errno`.
+/// It's best to make one read attempt to the network per call. Additional reads will
+/// be triggered by subsequent calls to one of the `_read_tls` methods.
+/// `userdata` is set to the value provided to `rustls_*_session_set_userdata`. In most
+/// cases that should be a struct that contains, at a minimum, a file descriptor.
+/// The buf and out_n pointers are borrowed and should not be retained across calls.
+pub type rustls_read_callback = Option<
+    unsafe extern "C" fn(
+        userdata: *mut c_void,
+        buf: *mut u8,
+        n: size_t,
+        out_n: *mut usize,
+    ) -> rustls_io_error,
+>;
+
+pub(crate) type ReadCallback = unsafe extern "C" fn(
+    userdata: *mut c_void,
+    buf: *mut u8,
+    n: size_t,
+    out_n: *mut usize,
+) -> rustls_io_error;
+
+pub(crate) struct CallbackReader {
+    pub callback: ReadCallback,
+    pub userdata: *mut c_void,
+}
+
+impl Read for CallbackReader {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let mut out_n: usize = 0;
+        let cb = self.callback;
+        let result = unsafe { cb(self.userdata, buf.as_mut_ptr(), buf.len(), &mut out_n) };
```

Comment:
I'm curious -- why assign `self.callback` to `cb` here? Why not just call `self.callback(self.userdata, ...)` in the `unsafe{}`?

Reply:
`self.callback(self.userdata, ...)` is ambiguous between method call and field access. So I'd have to write it as `(self.callback)(self.userdata, ...)`, which I like less.

Reply:
Oh yes, I've run into that and it's quite annoying. Your version is indeed better.